xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py

```diff
@@ -0,0 +1,607 @@
+        else:
+            raise Exception("args for IBlock ill structured")
+
+        object.__setattr__(self, "args", IList(block_args))
+        object.__setattr__(self, "ops", IList(ops))
+
+        self.args.freeze()
+        self.ops.freeze()
+
+    @classmethod
+    def from_mutable(
+        cls,
+        block: Block,
+        value_map: dict[SSAValue, ISSAValue] | None = None,
+        block_map: dict[Block, IBlock] | None = None,
+    ) -> IBlock:
+        """
+        Creates an immutable block from a mutable block. 
+        The value_map and block_map are used to map already known correspondings
+        of mutable values to immutable values and mutable blocks to immutable blocks.
+        """
+        if value_map is None:
+            value_map = {}
+        if block_map is None:
+            block_map = {}
+
+        args: list[IBlockArg] = []
+        for arg in block.args:
+            # The IBlock that will house this IBlockArg is not constructed yet.
+            # After construction the block field will be set by the IBlock.
+            immutable_arg = IBlockArg(
+                arg.typ,
+                IList([]),
+                None,  # type: ignore
+                arg.index)
+            args.append(immutable_arg)
+            value_map[arg] = immutable_arg
+
+        immutable_ops = [
+            IOperation.from_mutable(op,
+                                    value_map=value_map,
+                                    block_map=block_map,
+                                    existing_operands=None) for op in block.ops
+        ]
+
+        return IBlock(args, immutable_ops)
+
+    def get_mutable_copy(
+            self,
+            value_mapping: dict[ISSAValue, SSAValue] | None = None,
+            block_mapping: dict[IBlock, Block] | None = None) -> Block:
+        """
+        Returns a mutable block that is a copy of this immutable block.
+        The value_mapping and block_mapping are used to map already known correspondings
+        of immutable values to mutable values and immutable blocks to mutable blocks.
+        """
+        if value_mapping is None:
+            value_mapping = {}
+        if block_mapping is None:
+            block_mapping = {}
+
+        # Block might already have been created by the Region, look it up
+        if self in block_mapping:
+            mutable_block = block_mapping[self]
+        else:
+            mutable_block = Block.from_arg_types(self.arg_types)
+        for idx, arg in enumerate(self.args):
+            value_mapping[arg] = mutable_block.args[idx]
+        block_mapping[self] = mutable_block
+
+        for immutable_op in self.ops:
+            mutable_block.add_op(
+                immutable_op.get_mutable_copy(value_mapping=value_mapping,
+                                              block_mapping=block_mapping))
+        return mutable_block
+
+
+def get_immutable_copy(op: Operation) -> IOperation:
```

Comment:
Shouldn't this be part of the (mutable) Operation class? Feels like it would make more sense there.

Reply:
I agree this makes more sense there.
The problem is we then create a circular import dependency, which I am not sure how to best break.  Do you have an idea around this?

Reply:
The classic solution is inline imports, I think, but maybe leave as is for now and refactor later :+1: